More edits.

richsalz opened the pull request:
Required the endpoitn array to be all of the same type: alias or service entries.  Do we want that???  Wording in the text is still contradictory, so we have to make up our minds :)

bemasc commented:
At MUST level (syntactically), SVCB allows a mix of AliasMode and ServiceMode records.  At SHOULD level (i.e. within present semantics), it allows either ServiceMode records or _one_ AliasMode record.  I think we should probably pick one or the other of those.  Allowing multiple AliasMode records but not a mix seems like an odd middle ground.

I think there's a reasonable argument to be made that if we relax the semantic rules in the future, it would be good to be able to use the existing formats for WKECH, and maybe even the existing translators.  That argues in favor of flexibility.

richsalz commented:
I prefer a mix of Alias/Service records and agree about the flexibility and will post a fixup commit tomorrow.

If CDN-A has one ECH key for all its customers and CDN-B has one ECH key for all its customers and the origin is a customer of both CDN's, then you could use multiple Alias records, no?

bemasc commented:
> If CDN-A has one ECH key for all its customers and CDN-B has one ECH key for all its customers and the origin is a customer of both CDN's, then you could use multiple Alias records, no?

As specified, SVCB/HTTPS doesn't allow that.  Instead, you need something like WKECH, so you can combine A and B's ServiceMode RRSets, and deliver the combination to your ZF with continuous updates.

richsalz commented:
Right, oops, forgot about only one AliasMode.  So I think the example of two alias entries should be fixed and you can then have
- Zero length array
- Array of length one (either alias or service entry)
- Array of length >1 (all service entries)

thoughts?

bemasc commented:
It sounds like you're saying that we should let the format be flexible, but stick to examples that comply with all the SVCB SHOULDs.  That seems like a good plan to me.

richsalz commented:
Fixup commit pushed, please take a look.

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+        of the entries:
+    <list style="symbols">
+      <t>
+        An empty endpoints array corresponds to an HTTPS record with inferred
+        SvcPriority, the TargetName equal to ".", and no ECH support.  This can
+        can be useful as a simple way to make use of the HTTPS RR
```
I don't think this is what we want.  An empty array should mean "delete all HTTPS records".  An array containing an empty object should mean this.

richsalz replied:
Makes sense; added a new commit to do this.

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+        be resolved.
+      </t>
+      <t>
+	If the endpoints array has more than one element, every item MUST
+        be a ServiceMode entry.
```
```suggestion
	If the endpoints array has more than one element, every item SHOULD
        be a ServiceMode entry, due to restrictions on the use of multiple AliasMode
        records (<relref target="RFC9460" section="2.4.2" displayFormat="comma" />).
```

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
-        the ZF MUST NOT update the DNS. Such failures will not be directly visible to
-        the client-facing server, so ZF implementations will need to provide some
-        form of reporting so that the situation can be resolved.
-        Note that this can lead to inconsistent behavior for a single origin served by multiple ZFs.
+	The intent of these restrictions is to reduce the chance of misconfiguration
+	by the ZF and to avoid introducing semantics not defined by
+        <xref target="RFC9460"/>.
     </t>
```
With the change from MUST to SHOULD, I'm not sure there's any restriction left for the ZF to enforce.

bemasc replied:
The draft defines a format, but it doesn't specify many "restrictions" (although there are a few, like no negative priorities, no weird characters in "target").  I've suggested new text for this paragraph.

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+    <t>The following keys are defined for both types of entries:
+    <list style="symbols">
+      <t>
+	  regeninterval: the number of seconds between key generation
+	  actions at the origin, i.e. a replacement ECHConfigList may be
+	  generated this often.
```
`regeninterval` is an attribute of the outer object, not the objects in the array.

richsalz replied:
That's not what the samples show.

bemasc replied:
OK, let's fix that.  The top-level object should have two keys: "endpoints" and "regeninterval".

richsalz replied:
Making regeninterval top-level fields have a two-part fix: here and in #29

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+	  params: A JSON Dictionary representing the SVCB SvcParams.  Each
+	  key in the dictionary is a string containing a registered
+	  SvcParamKey name (e.g., "ipv6hint") or a SvcParamKey in generic
+	  form (e.g., "key65528").
```
```suggestion
	  form (e.g., "key65528").  The default value is "{}".
```

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+
+    <list style="symbols">
+      <t>
+	  alias: The value MUST be a DNS name that could be used as the
+	  TargetName of an HTTPS resource record.  This indicates that the
```
I think this name should have the same character restrictions as `target`, whatever those may be.  (RFC 9460 imposes no restrictions on either, but there is a [related interesting wrinkle](https://datatracker.ietf.org/doc/html/rfc9460#section-10.3).)

richsalz replied:
Maybe "same restrictions as `target` but see section 10 of 9460" ?

bemasc replied:
That section applies equally to `target` and `alias`.  I probably just wouldn't mention it in this document.

richsalz replied:
okay, so removed the charset section in that "target" entry up at line 403.

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+	  this directive by publishing an AliasMode record, publishing a
+	  CNAME record, copying HTTPS records from the target zone, or
+	  fetching https://cfs.example.com/.well-known/origin-svcb (if it
+	  exists).  In this case, the regeninterval indicates that backend
+	  does not plan to change the content at the URL for at least that
+	  number of seconds.  If an alias entry is present then any ech and
+	  port entries (if also present) MUST be ignored.
```
```suggestion
	  exists).
```

I think "regeninterval" should consistently mean "how often this JSON file is updated", and the backend can make its own decisions about how often to update its configuration.

I don't think the "ech" and "port" rule applies anymore.

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -345,120 +342,147 @@ Client <-->|                    | |                    | facing
+  <t>These definitions, taken with the <xref target="zfbehave">ZF behaviour</xref>
+      specified below, provide the following important properties:
+      <list style="symbols">
+	<t>
+	    Origins can indicate that different ECHConfigs are used on
+	    different ports.
+	</t>
+	<t>
+	    Origins can indicate that multiple CDNs are in use, each with
+	    its own ECHConfig.
+	</t>
+	<t>
+	    Origins that simply alias to a single target can indicate this
+	    without copying the ECHConfig and other parameters, which
+	    can interfere with key rotation and other maintenance.
+	</t>
+	<t>
+	    "port" and "target" are generally sufficient to uniquely identify
+	    a ServiceMode record, so zone factories can use the endpoint list
+	    to add ECH to pre-existing ServiceMode records that may have other
+	    SvcParams.
+	</t>
```
```suggestion
	<t>
	    Origins can express any useful configuration that is representable by HTTPS records, including multiple endpoints representing different ports, providers, etc.
	</t>
	<t>
	    Origins that simply alias to a single target can indicate this
	    without copying the ECHConfig and other parameters, avoiding
	    the maintenance burden of staying synchronized with the target's
	    key rotations and configuration updates.
	</t>
```

richsalz commented:
Do we want to insist on one regen interval for all ServiceMode entries?  Different CDNs, for example, might regen keys differently.  I'm fine with moving regeninterval to top-level as long as we're making an informed decision.

bemasc commented:
`regeninterval` serves two purposes:
 * It tells the ZF how frequently to poll for updates.
 * It hints to the ZF how long a DNS TTL it should use.

Polling is necessarily requesting the whole JSON file, and the DNS TTL applies to the whole HTTPS RRset (all RRs must have the same TTL).  If there were multiple `regeninterval`s, the ZF would have to combine them somehow into a single number anyway.  I think the origin is better placed to do that, since it knows where the contents are coming from.

richsalz commented:
I made reginterval a top-level field, and pulled it out of the individual items in the `endpoints` array. The other part of that change is updating the IANA section, which will be in #29 shortly.  But I believe I have addressed all your concerns (so far:) here.

bemasc commented on draft-ietf-tls-wkech.xml:
```diff
@@ -332,133 +331,159 @@ Client <-->|                    | |                    | facing
+	  domain name, corresponding to the HTTPS record's TargetName.
+	  The default value is ".".
+      </t>
+      <t>
```
```suggestion
      <t>
           [[ TODO: Restrict the allowed characters or specify the encoding rules for TargetName. ]]
      </t>
      <t>
```

richsalz commented:
Merged by hand on my machine; too many merge conflicts. Closing this PR
